### PR TITLE
[PhpUnitBridge] When using phpenv + phpenv-composer plugin, composer executable is wrapped into a bash script

### PR DESCRIPTION
**Pull request by oleg-andreyev**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | #34796
| License       | MIT
| Doc PR        | -

Locally using [phpenv](https://github.com/phpenv/phpenv) together [phpenv-composer](https://github.com/ngyuki/phpenv-composer) with for managing php versions and composer installations.

`which composer` is detecting wrapper instead of binary.
Before appending PHP executable to composer path need to validate that detected composer is not a wrapper otherwise PHP will simply return the content of a wrapper

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
What do you think of https://github.com/symfony/symfony/pull/35166#issuecomment-570778609?

**Comment by oleg-andreyev** [comment]
@nicolas-grekas while `phpenv which composer` will work for detecting `phpenv-composer` proxy, it's still a bash script and I think that it's still possible that developers may wrap composer executable with custom wrapper.

About COMPOSER_BIN from https://github.com/symfony/symfony/pull/35166 
IMO it's better to add detection instead of env. variable.

File: src/Symfony/Bridge/PhpUnit/bin/simple-phpunit [hunk 1]
```diff
@@ -57,10 +57,11 @@ foreach ($defaultEnvs as $envName => $envValue) {
...
 }
 
+$isWindows = '\\' === DIRECTORY_SEPARATOR;
 $COMPOSER = file_exists($COMPOSER = $oldPwd.'/composer.phar')
-    || ($COMPOSER = rtrim('\\' === DIRECTORY_SEPARATOR ? preg_replace('/[\r\n].*/', '', `where.exe composer.phar`) : `which composer.phar 2> /dev/null`))
-    || ($COMPOSER = rtrim('\\' === DIRECTORY_SEPARATOR ? preg_replace('/[\r\n].*/', '', `where.exe composer`) : `which composer 2> /dev/null`))
-    ? $PHP.' '.escapeshellarg($COMPOSER)
+    || ($COMPOSER = rtrim($isWindows ? preg_replace('/[\r\n].*/', '', `where.exe composer.phar`) : `which composer.phar 2> /dev/null`))
+    || ($COMPOSER = rtrim($isWindows ? preg_replace('/[\r\n].*/', '', `where.exe composer`) : `which composer 2> /dev/null`))
+    ? (!$isWindows && (strpos(`file $COMPOSER 2> /dev/null`, 'binary data') === false) ? '' : $PHP).' '.escapeshellarg($COMPOSER)
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/bin/simple-phpunit]: instead, I'd suggest checking if the file starts by `#!/usr/bin/env php`, both on windows and on linux

File: src/Symfony/Bridge/PhpUnit/bin/simple-phpunit [hunk 2]
```diff
@@ -57,10 +57,11 @@ foreach ($defaultEnvs as $envName => $envValue) {
     }
 }
 
+$isWindows = '\\' === DIRECTORY_SEPARATOR;
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/bin/simple-phpunit]: please revet this change: it's unrelated and it slows down reviewing

File: src/Symfony/Bridge/PhpUnit/bin/simple-phpunit (hunk 1 above)
Comment (oleg-andreyev) [reply src/Symfony/Bridge/PhpUnit/bin/simple-phpunit]: wrapped does not have to be a php file.
Example `$ cat phpenv which composer`

```bash
$ cat `phpenv which composer`

#!/usr/bin/env bash
# Summary: phpenv composer plugin

set -e

prefix=$(phpenv prefix)
...
```

File: src/Symfony/Bridge/PhpUnit/bin/simple-phpunit (hunk 2 above)
Comment (oleg-andreyev) [reply src/Symfony/Bridge/PhpUnit/bin/simple-phpunit]: 👌

File: src/Symfony/Bridge/PhpUnit/bin/simple-phpunit (hunk 1 above)
Comment (oleg-andreyev) [reply src/Symfony/Bridge/PhpUnit/bin/simple-phpunit]: There could be at least 2 different shebangs:
```
#!interpreter [optional-arg]

#!/bin/sh (bash or any other shell)
#!/usr/bin/env python (php bash sh and etc.)
``` 

Need to avoid appending php-executable for at least following cases:
```
#!/usr/bin/env php
#!<anything_path_to>php
#!/bin/sh
#!/bin/bash
```

so I propose to not to append php-executable if any shebang `#!`  is detected.

Comment (nicolas-grekas) [reply src/Symfony/Bridge/PhpUnit/bin/simple-phpunit]: Works for me.

Comment (oleg-andreyev) [reply src/Symfony/Bridge/PhpUnit/bin/simple-phpunit]: updated PR.

Comment (oleg-andreyev) [reply src/Symfony/Bridge/PhpUnit/bin/simple-phpunit]: Updated PR

File: src/Symfony/Bridge/PhpUnit/bin/simple-phpunit [hunk 3]
```diff
@@ -57,10 +57,11 @@ foreach ($defaultEnvs as $envName => $envValue) {
     }
 }
 
-$COMPOSER = file_exists($COMPOSER = $oldPwd.'/composer.phar')
+$COMPOSER =
+    file_exists($COMPOSER = $oldPwd.'/composer.phar')
     || ($COMPOSER = rtrim('\\' === DIRECTORY_SEPARATOR ? preg_replace('/[\r\n].*/', '', `where.exe composer.phar`) : `which composer.phar 2> /dev/null`))
     || ($COMPOSER = rtrim('\\' === DIRECTORY_SEPARATOR ? preg_replace('/[\r\n].*/', '', `where.exe composer`) : `which composer 2> /dev/null`))
-    ? $PHP.' '.escapeshellarg($COMPOSER)
+    ? (!'\\' === DIRECTORY_SEPARATOR && (fread($fp = fopen($COMPOSER, 'rb'), 2) === '#!' && fclose($fp)) ? '' : $PHP).' '.escapeshellarg($COMPOSER) // checking that detected COMPOSER has shebang at the start, if it's found do not append php-executable
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/bin/simple-phpunit]: I think this would work better:
`? (file_get_contents($COMPOSER, false, null, 0, 18) === '#!/usr/bin/env php' ? $PHP : '').' '.escapeshellarg($COMPOSER) // detect shell wrappers by looking at the shebang`
this works on linux and windows, and detects php exactly. This shebang is the one that composer uses since always

File: src/Symfony/Bridge/PhpUnit/bin/simple-phpunit [hunk 4]
```diff
@@ -57,10 +57,11 @@ foreach ($defaultEnvs as $envName => $envValue) {
     }
 }
 
-$COMPOSER = file_exists($COMPOSER = $oldPwd.'/composer.phar')
+$COMPOSER =
+    file_exists($COMPOSER = $oldPwd.'/composer.phar')
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/bin/simple-phpunit]: please revert this change :)

File: src/Symfony/Bridge/PhpUnit/bin/simple-phpunit (hunk 3 above)
Comment (oleg-andreyev) [reply src/Symfony/Bridge/PhpUnit/bin/simple-phpunit]: Shame to me 😟 , I didn't know that `file_get_contents` could read file with offset.

Comment (nicolas-grekas) [reply src/Symfony/Bridge/PhpUnit/bin/simple-phpunit]: no worries, I copy/pasted from SO :)

File: src/Symfony/Bridge/PhpUnit/bin/simple-phpunit (hunk 4 above)
Comment (oleg-andreyev) [reply src/Symfony/Bridge/PhpUnit/bin/simple-phpunit]: done

File: src/Symfony/Bridge/PhpUnit/bin/simple-phpunit (hunk 3 above)
Comment (oleg-andreyev) [reply src/Symfony/Bridge/PhpUnit/bin/simple-phpunit]: done

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @oleg-andreyev.
